src/main/java/org/jabref/logic/texparser/DefaultTexParser.java

```diff
@@ -0,0 +1,141 @@
+package org.jabref.logic.texparser;
+
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jabref.model.texparser.Citation;
+import org.jabref.model.texparser.TexParser;
+import org.jabref.model.texparser.TexParserResult;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultTexParser implements TexParser {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultTexParser.class);
+
+    /**
+     * It is allowed to add new cite commands for pattern matching.
+     *
+     * <p>Some valid examples: "citep", "[cC]ite", "[cC]ite(author|title|year|t|p)?"
+     *
+     * <p>TODO: Add support for multicite commands.
+     */
+    private static final String[] CITE_COMMANDS = new String[] {
+            "[cC]ite(alt|alp|author|authorfull|date|num|p|t|text|title|url|year|yearpar)?",
+            "([aA]|fnote|foot|footfull|full|no|[nN]ote|[pP]aren|[pP]note|[tT]ext|[sS]mart|super)cite",
+            "footcitetext"
+    };
+    private static final String CITE_REGEX = String.format("\\\\(?:%s)\\*?(?:\\[(?:[^\\]]*)\\]){0,2}\\{(?<key>[^\\}]*)\\}",
+            String.join("|", CITE_COMMANDS));
+
+    private static final String INCLUDE_REGEX = "\\\\(?:include|input)\\{(?<file>[^\\}]*)\\}";
+
+    private final TexParserResult result;
+
+    public DefaultTexParser() {
+        this.result = new TexParserResult();
+    }
+
+    @Override
+    public TexParserResult parse(String citeString) {
+        matchCitation(Paths.get("foo/bar"), 1, citeString);
+        return result;
+    }
+
+    @Override
+    public TexParserResult parse(Path texFile) {
+        return parse(Collections.singletonList(texFile));
+    }
+
+    @Override
+    public TexParserResult parse(List<Path> texFiles) {
+        List<Path> referencedFiles = new ArrayList<>();
+
+        if (result.getFileList().isEmpty()) {
+            result.getFileList().addAll(texFiles);
+        } else {
+            result.getNestedFiles().addAll(texFiles);
+        }
+
+        for (int fileIndex = 0; fileIndex < texFiles.size(); fileIndex++) {
```

Comment:
From what I see this could be converted to the "normal" advanced for. eg. for(Path file: textFiles) or do you really need the index(position information somewhere?
The advanced for, or "for each" effectively reduces the problem of IndexOutOfBounds exceptions